File: src/prototype/web/src/state/counter_window.ts
// PURE — defensive counter window per docs/design/input_system_defense_v1.md §D.
//
// The counter window mirrors the attacker's JudgmentWindowFSM lifecycle but
// with its own candidate set (counter techniques only). A counter window
// only opens in the same frame an attacker window enters OPENING for a
// technique that has a registered counter.
//
// Success effects (§D.2):
//   SCISSOR_COUNTER → forces the attacker's window to CLOSING (DISRUPTED)
//   TRIANGLE_EARLY_STACK → same + resets top.arm_extracted both sides to false

import {
  TIME_SCALE,
  WINDOW_TIMING,
  type Technique,
} from "./judgment_window.js";

export type CounterTechnique = "SCISSOR_COUNTER" | "TRIANGLE_EARLY_STACK";

// §D.2 — counter table. The attacker's technique → the defender's
// available counter. Not every technique has a counter in M1.
export const COUNTER_FOR: Readonly<Partial<Record<Technique, CounterTechnique>>> =
  Object.freeze({
    SCISSOR_SWEEP: "SCISSOR_COUNTER",
    TRIANGLE: "TRIANGLE_EARLY_STACK",
  });

export type CounterWindowState = "CLOSED" | "OPENING" | "OPEN" | "CLOSING";

export type CounterWindow = Readonly<{
  state: CounterWindowState;
  stateEnteredMs: number;
  candidates: ReadonlyArray<CounterTechnique>;
  cooldownUntilMs: number;
}>;

export const INITIAL_COUNTER_WINDOW: CounterWindow = Object.freeze({
  state: "CLOSED" as const,
  stateEnteredMs: Number.NEGATIVE_INFINITY,
  candidates: Object.freeze([]),
  cooldownUntilMs: Number.NEGATIVE_INFINITY,
});

// Used by stepSimulation: when the attacker's window enters OPENING, look
// up counter candidates for each firing technique. Callers pass the
// attacker candidate list; this function returns the filtered counter
// list (empty if nothing has a counter).
export function counterCandidatesFor(
  attackerCandidates: ReadonlyArray<Technique>,
): ReadonlyArray<CounterTechnique> {
  const out: CounterTechnique[] = [];
  for (const t of attackerCandidates) {
    const c = COUNTER_FOR[t];
    if (c !== undefined && !out.includes(c)) out.push(c);
  }
  return Object.freeze(out);
}

export type CounterTickInput = Readonly<{
  nowMs: number;
  openAttackerWindow: boolean;           // true during frames where the attacker window is OPENING or OPEN
  openingSeed: ReadonlyArray<CounterTechnique>; // set at the frame we should enter OPENING (one-shot)
  confirmedCounter: CounterTechnique | null;
  dismissRequested: boolean;
}>;

export type CounterTickEvent =
  | { kind: "COUNTER_WINDOW_OPENING"; candidates: ReadonlyArray<CounterTechnique> }
  | { kind: "COUNTER_WINDOW_OPEN" }
  | { kind: "COUNTER_WINDOW_CLOSING"; reason: CounterCloseReason }
  | { kind: "COUNTER_WINDOW_CLOSED" }
  | { kind: "COUNTER_CONFIRMED"; counter: CounterTechnique };

export type CounterCloseReason =
  | "CONFIRMED"
  | "DISMISSED"
  | "TIMED_OUT"
  | "ATTACKER_CLOSED"; // attacker window shut for any reason — defender loses the chance

export function tickCounterWindow(
  prev: CounterWindow,
  input: CounterTickInput,
  timing = WINDOW_TIMING,
): { next: CounterWindow; events: readonly CounterTickEvent[]; timeScale: number } {
  const events: CounterTickEvent[] = [];
  let next = prev;
  let timeScale = TIME_SCALE.normal;

  switch (prev.state) {
    case "CLOSED": {
      // Open only if the attacker's window just entered OPENING AND there
      // is at least one counter candidate. Cooldown honoured.
      if (
        input.openingSeed.length > 0 &&
        input.nowMs >= prev.cooldownUntilMs
      ) {
        next = Object.freeze({
          state: "OPENING" as const,
          stateEnteredMs: input.nowMs,
          candidates: Object.freeze([...input.openingSeed]),
          cooldownUntilMs: prev.cooldownUntilMs,
        });
        events.push({
          kind: "COUNTER_WINDOW_OPENING",
          candidates: next.candidates,
        });
      }
      break;
    }

    case "OPENING": {
      const t = (input.nowMs - prev.stateEnteredMs) / timing.openingMs;
      timeScale = lerp(TIME_SCALE.normal, TIME_SCALE.open, clamp01(t));
      if (!input.openAttackerWindow) {
        // Attacker vanished before we fully opened — bail.
        next = enterClosing(prev, input.nowMs);
        events.push({ kind: "COUNTER_WINDOW_CLOSING", reason: "ATTACKER_CLOSED" });
        break;
      }
      if (t >= 1) {
        next = Object.freeze({ ...prev, state: "OPEN" as const, stateEnteredMs: input.nowMs });
        events.push({ kind: "COUNTER_WINDOW_OPEN" });
        timeScale = TIME_SCALE.open;
      }
      break;
    }

    case "OPEN": {
      timeScale = TIME_SCALE.open;

      if (input.confirmedCounter !== null && prev.candidates.includes(input.confirmedCounter)) {
        events.push({ kind: "COUNTER_CONFIRMED", counter: input.confirmedCounter });
        next = enterClosing(prev, input.nowMs);
        events.push({ kind: "COUNTER_WINDOW_CLOSING", reason: "CONFIRMED" });
        break;
      }

      if (input.dismissRequested) {
        next = enterClosing(prev, input.nowMs);
        events.push({ kind: "COUNTER_WINDOW_CLOSING", reason: "DISMISSED" });
        break;
      }

      // Attacker's window collapsed — we lose our chance.
      if (!input.openAttackerWindow) {
        next = enterClosing(prev, input.nowMs);
        events.push({ kind: "COUNTER_WINDOW_CLOSING", reason: "ATTACKER_CLOSED" });
        break;
      }

      if (input.nowMs - prev.stateEnteredMs >= timing.openMaxMs) {
        next = enterClosing(prev, input.nowMs);
        events.push({ kind: "COUNTER_WINDOW_CLOSING", reason: "TIMED_OUT" });
      }
      break;
    }

    case "CLOSING": {
      const t = (input.nowMs - prev.stateEnteredMs) / timing.closingMs;
      timeScale = lerp(TIME_SCALE.open, TIME_SCALE.normal, clamp01(t));
      if (t >= 1) {
        next = Object.freeze({
          state: "CLOSED" as const,
          stateEnteredMs: input.nowMs,
          candidates: Object.freeze([]),
          cooldownUntilMs: input.nowMs + timing.cooldownMs,
        });
        events.push({ kind: "COUNTER_WINDOW_CLOSED" });
        timeScale = TIME_SCALE.normal;
      }
      break;
    }
  }

  return { next, events: Object.freeze(events), timeScale };
}

function enterClosing(prev: CounterWindow, nowMs: number): CounterWindow {
  return Object.freeze({ ...prev, state: "CLOSING" as const, stateEnteredMs: nowMs });
}

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

function clamp01(x: number): number {
  return x < 0 ? 0 : x > 1 ? 1 : x;
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/State/CounterWindow.cs
// Ported 1:1 from src/prototype/web/src/state/counter_window.ts.
// PURE — defensive counter window per docs/design/input_system_defense_v1.md §D.
//
// A counter window opens only when the attacker's JudgmentWindow enters OPENING
// for a technique that has a registered counter.
//
// Success effects (§D.2):
//   SCISSOR_COUNTER       → forces attacker window to CLOSING (DISRUPTED)
//   TRIANGLE_EARLY_STACK  → same + resets top.arm_extracted both sides to false

using System.Collections.Generic;

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // Counter techniques (§D.2)
    // -------------------------------------------------------------------------

    public enum CounterTechnique
    {
        ScissorCounter,
        TriangleEarlyStack,
    }

    public enum CounterWindowState { Closed, Opening, Open, Closing }

    public enum CounterCloseReason
    {
        Confirmed,
        Dismissed,
        TimedOut,
        AttackerClosed, // attacker window shut — defender loses the chance
    }

    // -------------------------------------------------------------------------
    // State
    // -------------------------------------------------------------------------

    public struct CounterWindow
    {
        public CounterWindowState  State;
        public long                StateEnteredMs;
        public CounterTechnique[]  Candidates;
        public long                CooldownUntilMs;

        public static readonly CounterWindow Initial = new CounterWindow
        {
            State           = CounterWindowState.Closed,
            StateEnteredMs  = BJJConst.SentinelTimeMs,
            Candidates      = System.Array.Empty<CounterTechnique>(),
            CooldownUntilMs = BJJConst.SentinelTimeMs,
        };
    }

    // -------------------------------------------------------------------------
    // Events
    // -------------------------------------------------------------------------

    public enum CounterEventKind
    {
        CounterWindowOpening,
        CounterWindowOpen,
        CounterWindowClosing,
        CounterWindowClosed,
        CounterConfirmed,
    }

    public struct CounterTickEvent
    {
        public CounterEventKind    Kind;
        public CounterTechnique[]  Candidates;     // CounterWindowOpening
        public CounterCloseReason  CloseReason;    // CounterWindowClosing
        public CounterTechnique    ConfirmedCounter; // CounterConfirmed
    }

    // -------------------------------------------------------------------------
    // Tick input
    // -------------------------------------------------------------------------

    public struct CounterTickInput
    {
        public long               NowMs;
        public bool               OpenAttackerWindow;  // true while attacker window is OPENING or OPEN
        public CounterTechnique[] OpeningSeed;         // set exactly on the frame we should enter OPENING
        public CounterTechnique?  ConfirmedCounter;    // nullable
        public bool               DismissRequested;
    }

    // -------------------------------------------------------------------------
    // Pure operations
    // -------------------------------------------------------------------------

    public static class CounterWindowOps
    {
        // §D.2 — counter table. Not every technique has a counter in M1.
        public static CounterTechnique[] CounterCandidatesFor(Technique[] attackerCandidates)
        {
            var result = new List<CounterTechnique>();
            foreach (var t in attackerCandidates)
            {
                if (TryGetCounter(t, out var c) && !result.Contains(c))
                    result.Add(c);
            }
            return result.ToArray();
        }

        static bool TryGetCounter(Technique t, out CounterTechnique c)
        {
            switch (t)
            {
                case Technique.ScissorSweep: c = CounterTechnique.ScissorCounter;      return true;
                case Technique.Triangle:     c = CounterTechnique.TriangleEarlyStack;  return true;
                default:                     c = default;                               return false;
            }
        }

        /// <summary>
        /// Advance the CounterWindow FSM by one tick.
        /// <paramref name="events"/> is append-only.
        /// </summary>
        public static CounterWindow Tick(
            CounterWindow prev,
            CounterTickInput input,
            List<CounterTickEvent> events,
            out float timeScale,
            JudgmentWindowTiming timing = default)
        {
            if (timing.OpeningMs == 0) timing = JudgmentWindowTiming.Default;

            timeScale = JudgmentWindowTimeScale.Normal;
            var next = prev;

            switch (prev.State)
            {
                case CounterWindowState.Closed:
                {
                    if (input.OpeningSeed != null && input.OpeningSeed.Length > 0 &&
                        input.NowMs >= prev.CooldownUntilMs)
                    {
                        next = new CounterWindow
                        {
                            State           = CounterWindowState.Opening,
                            StateEnteredMs  = input.NowMs,
                            Candidates      = CopyArray(input.OpeningSeed),
                            CooldownUntilMs = prev.CooldownUntilMs,
                        };
                        events.Add(new CounterTickEvent
                        {
                            Kind       = CounterEventKind.CounterWindowOpening,
                            Candidates = next.Candidates,
                        });
                    }
                    break;
                }

                case CounterWindowState.Opening:
                {
                    float t = (float)(input.NowMs - prev.StateEnteredMs) / timing.OpeningMs;
                    timeScale = Lerp(JudgmentWindowTimeScale.Normal, JudgmentWindowTimeScale.Open, Clamp01(t));

                    if (!input.OpenAttackerWindow)
                    {
                        next = EnterClosing(prev, input.NowMs);
                        events.Add(new CounterTickEvent
                        {
                            Kind         = CounterEventKind.CounterWindowClosing,
                            CloseReason  = CounterCloseReason.AttackerClosed,
                        });
                        break;
                    }
                    if (t >= 1f)
                    {
                        next = new CounterWindow
                        {
                            State           = CounterWindowState.Open,
                            StateEnteredMs  = input.NowMs,
                            Candidates      = prev.Candidates,
                            CooldownUntilMs = prev.CooldownUntilMs,
                        };
                        events.Add(new CounterTickEvent { Kind = CounterEventKind.CounterWindowOpen });
                        timeScale = JudgmentWindowTimeScale.Open;
                    }
                    break;
                }

                case CounterWindowState.Open:
                {
                    timeScale = JudgmentWindowTimeScale.Open;

                    if (input.ConfirmedCounter.HasValue &&
                        ArrayContains(prev.Candidates, input.ConfirmedCounter.Value))
                    {
                        events.Add(new CounterTickEvent
                        {
                            Kind             = CounterEventKind.CounterConfirmed,
                            ConfirmedCounter = input.ConfirmedCounter.Value,
                        });
                        next = EnterClosing(prev, input.NowMs);
                        events.Add(new CounterTickEvent
                        {
                            Kind        = CounterEventKind.CounterWindowClosing,
                            CloseReason = CounterCloseReason.Confirmed,
                        });
                        break;
                    }

                    if (input.DismissRequested)
                    {
                        next = EnterClosing(prev, input.NowMs);
                        events.Add(new CounterTickEvent
                        {
                            Kind        = CounterEventKind.CounterWindowClosing,
                            CloseReason = CounterCloseReason.Dismissed,
                        });
                        break;
                    }

                    if (!input.OpenAttackerWindow)
                    {
                        next = EnterClosing(prev, input.NowMs);
                        events.Add(new CounterTickEvent
                        {
                            Kind        = CounterEventKind.CounterWindowClosing,
                            CloseReason = CounterCloseReason.AttackerClosed,
                        });
                        break;
                    }

                    if (input.NowMs - prev.StateEnteredMs >= timing.OpenMaxMs)
                    {
                        next = EnterClosing(prev, input.NowMs);
                        events.Add(new CounterTickEvent
                        {
                            Kind        = CounterEventKind.CounterWindowClosing,
                            CloseReason = CounterCloseReason.TimedOut,
                        });
                    }
                    break;
                }

                case CounterWindowState.Closing:
                {
                    float t = (float)(input.NowMs - prev.StateEnteredMs) / timing.ClosingMs;
                    timeScale = Lerp(JudgmentWindowTimeScale.Open, JudgmentWindowTimeScale.Normal, Clamp01(t));
                    if (t >= 1f)
                    {
                        next = new CounterWindow
                        {
                            State           = CounterWindowState.Closed,
                            StateEnteredMs  = input.NowMs,
                            Candidates      = System.Array.Empty<CounterTechnique>(),
                            CooldownUntilMs = input.NowMs + timing.CooldownMs,
                        };
                        events.Add(new CounterTickEvent { Kind = CounterEventKind.CounterWindowClosed });
                        timeScale = JudgmentWindowTimeScale.Normal;
                    }
                    break;
                }
            }

            return next;
        }

        // -----------------------------------------------------------------------

        static CounterWindow EnterClosing(CounterWindow prev, long nowMs) =>
            new CounterWindow
            {
                State           = CounterWindowState.Closing,
                StateEnteredMs  = nowMs,
                Candidates      = prev.Candidates,
                CooldownUntilMs = prev.CooldownUntilMs,
            };

        static bool ArrayContains(CounterTechnique[] arr, CounterTechnique t)
        {
            foreach (var x in arr) if (x == t) return true;
            return false;
        }

        static CounterTechnique[] CopyArray(CounterTechnique[] src)
        {
            var dst = new CounterTechnique[src.Length];
            System.Array.Copy(src, dst, src.Length);
            return dst;
        }

        static float Lerp(float a, float b, float t) => a + (b - a) * t;
        static float Clamp01(float v) => v < 0f ? 0f : v > 1f ? 1f : v;
    }
}
